docs/man/openvpn3-service-aws.8.rst
==========================
openvpn3-service-aws
==========================

----------------------
OpenVPN 3 Linux client
----------------------

:Manual section: 8
:Manual group: OpenVPN 3 Linux

SYNOPSIS
========
| ``openvpn3-service-aws`` ``[OPTIONS]``
| ``openvpn3-service-aws`` ``-h`` | ``--help``

DESCRIPTION
===========
The ``openvpn3-service-aws`` is a backend service for OpenVPN 3 Linux,
which enables hosts in VPN subnet to communicate to hosts in VPC subnet.
Upon start, it reads its configuration file and subscribes to
:code:`ROUTE_ADDED` and :code:`ROUTE_REMOVED` signals from
``openvpn3-service-netcfg`` service.  On receiving those signals, service
adds/deletes VPN routes to/from VPC subnet using Amazon EC2 API.

This service is normally started automatically on boot, which is achieved by enabling it
with provided **openvpn3-aws.service** *systemd* unit file.  Alternatively service could
be started manually on the command line.  This process should run as the *openvpn* user
and will automatically switch to this user account if it has the needed privileges to do so.
Beware that changing this to another user account also requires updating the D-Bus policy
for the ``net.openvpn.v3.aws`` service as well.

CONFIGURATION
=============
To use Amazon EC2 API, service uses temporary security credentials, which are retrieved from
instance metadata.  Credentials are tied to an IAM role, which must be assigned to the EC2
instance where the service is running.

Below is a part of a CloudFormation template, which creates role and instance profile:

::

    "RouteManagerRole": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Version": "2012-10-17",
          "Statement": [
            {
              "Effect": "Allow",
              "Principal": {
                "Service": [
                  "ec2.amazonaws.com"
                ]
              },
              "Action": [
                "sts:AssumeRole"
              ]
            }
          ]
        },
        "Policies": [
          {
            "PolicyName": "RouteManagerPolicy",
            "PolicyDocument": {
              "Version": "2012-10-17",
              "Statement": [
                {
                  "Effect": "Allow",
                  "Action": [
                    "ec2:CreateRoute",
                    "ec2:DescribeNetworkInterfaceAttribute",
                    "ec2:DescribeNetworkInterfaces",
                    "ec2:DescribeRouteTables",
                    "ec2:ModifyNetworkInterfaceAttribute",
                    "ec2:ReplaceRoute",
                    "ec2:DeleteRoute"
                  ],
                  "Resource": "*"
                },
                {
                  "Effect": "Allow",
                  "Action": [
                    "ec2messages:GetMessages",
                    "ssm:UpdateInstanceInformation",
                    "ssm:ListInstanceAssociations",
                    "ssm:ListAssociations",
                    "ssmmessages:CreateControlChannel",
                    "ssmmessages:CreateDataChannel",
                    "ssmmessages:OpenControlChannel",
                    "ssmmessages:OpenDataChannel"
                  ],
                  "Resource": "*"
                }
              ]
            }
          }
        ]
      }
    },
    "InstanceProfile": {
      "Type": "AWS::IAM::InstanceProfile",
      "Properties": {
        "Roles": [
          {
            "Ref": "RouteManagerRole"
          }
        ]
      }
    }

An instance profile must be associated with the instance:

::

  "Resources": {
    "EC2Instance": {
      "Type": "AWS::EC2::Instance",
      "Properties": {
        "IamInstanceProfile": {
          "Ref": "InstanceProfile"
        }
      }
    }
  }

In order to retrieve the credentials needed, the ``openvpn3-service-aws``
service requires a role name, which is read from the configuration file
(:code:`/etc/openvpn/openvpn3-aws.json`).  Here is a CloudFormation
snippet which creates this file:

::

  "Metadata": {
    "AWS::CloudFormation::Init": {
      "config": {
        "files": {
          "/etc/openvpn3/openvpn3-aws.json": {
            "content": {
              "Fn::Join": [
                "",
                [
                  "{\n",
                  "\"role\": \"",
                  {
                    "Ref": "RouteManagerRole"
                  },
                  "\"\n",
                  "}\n"
                ]
              ]
            },
            "mode": "000644",
            "owner": "root",
            "group": "root"
           }
        }
      }
    }
  }

Below is an example of configuration file, generated by the script above:

::

  {
    "role": "lev1-RouteManagerRole-1QGQ76R9S4C11"
  }

OPTIONS
=======

-h, --help      Print  usage and help details to the terminal

--version       Prints the version of the program and exists

-c FILE, --config FILE
                Path of configuration file from where role name is read.
                The default is :code:`/etc/openvpn/openvpn3-aws.json`.

--log-level LEVEL
                Sets the default log verbosity for log events generated by
                this service.  The default is :code:`3`.  Valid values are
                :code:`0` to :code:`6`.  Higher log levels results in more
                verbose logs and log level :code:`6` will contain all debug
                log events.

--log-file LOG_DESTINATION
                By default, logging will go via the ``openvpn3-service-logger``
                service.  By providing this argument, logging will also be sent
                to *LOG_DESTINATION*, which can be either a filename or
                :code:`stdout:` where the latter one sends log data to the
                console.

--colour
                This will add colours to log events when logging to file
                or console.  Log events will be coloured based on the log
                level of the event.

--signal-broadcast
                Normally, the ``openvpn3-service-aws`` will attach a
                log stream to the ``openvpn3-service-logger`` service.  By
                providing this option, it will instead broadcast all log events
                to all listeners on the system.

SEE ALSO
========

``openvpn3-linux``\(7)
``openvpn3-service-netcfg``\(8)


addons/aws/openvpn3-service-aws.cpp
//  OpenVPN 3 Linux client -- Next generation OpenVPN client
//
//  Copyright (C) 2019 - 2020  OpenVPN Inc. <sales@openvpn.net>
//  Copyright (C) 2019 - 2020  Lev Stipakov <lev@openvpn.net>
//
//  This program is free software: you can redistribute it and/or modify
//  it under the terms of the GNU Affero General Public License as
//  published by the Free Software Foundation, version 3 of the
//  License.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU Affero General Public License for more details.
//
//  You should have received a copy of the GNU Affero General Public License
//  along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

#include <iostream>
#include <iomanip>
#include <sstream>
#include <exception>
#include <set>
#include <glib-unix.h>

// Needs to be included before openvpn3-core library
// It provides needed logging facility used by Core lib.
#include "log/core-dbus-logbase.hpp"

#include <openvpn/random/randapi.hpp>
#include <openvpn/ssl/sslchoose.hpp>
#include <openvpn/init/initprocess.hpp>
#include <openvpn/aws/awsroute.hpp>
#include <openvpn/common/jsonfile.hpp>

#include "common/utils.hpp"
#include "dbus/core.hpp"
#include "dbus/connection-creds.hpp"
#include "log/logwriter.hpp"
#include "log/ansicolours.hpp"
#include "log/proxy-log.hpp"
#include "netcfg/netcfg-changeevent.hpp"
#include "netcfg/proxy-netcfg.hpp"
#include "netcfg/netcfg-changeevent.hpp"


using namespace openvpn;

#define OPENVPN3_AWS_CONFIG "/etc/openvpn3/openvpn3-aws.json"
static const std::string OpenVPN3DBus_interf_aws = "net.openvpn.v3.aws";
static const std::string OpenVPN3DBus_path_aws = "/net/openvpn/v3/aws";


class AWSObject : public RC<thread_safe_refcount>,
                  public DBusObject,
                  public DBusSignalSubscription {
public:
    typedef  RCPtr<AWSObject> Ptr;

    AWSObject(GDBusConnection* conn, const std::string& objpath,
              const std::string& config_file,
              LogWriter* logwr, unsigned int log_level, bool signal_broadcast)
          : DBusObject(objpath),
            DBusSignalSubscription(conn, "", "", "", "NetworkChange"),
            netcfgmgr(conn),
            vpcRoutes {},
            log_sender(conn, LogGroup::EXTSERVICE,
                       OpenVPN3DBus_interf_aws, objpath, logwr)
    {
        log_sender.SetLogLevel(log_level);
        if (!signal_broadcast)
        {
            DBusConnectionCreds credsprx(conn);
            log_sender.AddTargetBusName(credsprx.GetUniqueBusID(OpenVPN3DBus_name_log));

        }
        std::stringstream introspection_xml;
        introspection_xml << "<node name='" << objpath << "'>"
                          << "    <interface name='"
                          << OpenVPN3DBus_interf_aws << "'>"
                          << "    </interface>" << "</node>";
        ParseIntrospectionXML(introspection_xml);

        // Retrieve AWS role credentials and retrieve needed VPC info
        role_name = read_role_name(config_file);
        if (role_name.empty())
        {
            THROW_DBUSEXCEPTION("AWSObject", "No role defined");
        }
        log_sender.LogInfo("Fetching credentials from role '"
                            + role_name + "'");

        auto route_context = prepare_route_context(role_name);
        AWS::Route::Info route_info { *route_context };
        route_table_id = route_info.route_table_id;
        network_interface_id = route_info.network_interface_id;
        AWS::Route::set_source_dest_check(*route_context,
                                          network_interface_id, false);

        log_sender.LogInfo("Running on instance " + route_context->instance_id()
                           + ", route table " + route_table_id);

        // We will act upon route changes caused by VPN sessions,
        // notifications which are sent by the net.openvpn.v3.netcfg service
        netcfgmgr.NotificationSubscribe(NetCfgChangeType::ROUTE_ADDED
                                        | NetCfgChangeType::ROUTE_REMOVED);
    }

    ~AWSObject() override
    {
        // We are shutting down, stop notification subscriptions
        // before we start cleaning up.
        netcfgmgr.NotificationUnsubscribe();

        // Retrieve AWS credentials and remove routes we are responsible
        // for from VPC
        auto route_context = prepare_route_context(role_name);

        for (auto vpcRoute : vpcRoutes)
        {
            try
            {
                AWS::Route::delete_route(*route_context,
                                         route_table_id,
                                         vpcRoute.first,
                                         vpcRoute.second);

                log_sender.LogInfo("Removed route " + vpcRoute.first);
            }
            catch (const std::exception& ex)
            {
                log_sender.LogError("Error removing route: " + std::string(ex.what()));
            }
        }

        vpcRoutes.clear();
    }

    /**
     *  Called each time the subscribed signal has a match on the D-Bus
     */
    void callback_signal_handler(GDBusConnection *connection,
                                 const std::string sender_name,
                                 const std::string object_path,
                                 const std::string interface_name,
                                 const std::string signal_name,
                                 GVariant *parameters) override
    {
        if (signal_name != "NetworkChange")
        {
            return;
        }

        // Parse the network change event, only consider route changes
        NetCfgChangeEvent ev(parameters);
        if ((ev.type != NetCfgChangeType::ROUTE_ADDED) &&
            (ev.type != NetCfgChangeType::ROUTE_REMOVED))
        {
            return;
        }

        try
        {
            const std::string cidr = ev.details["subnet"] + "/" + ev.details["prefix"];
            const bool ipv6 =  ev.details["ip_version"] == "6";
            auto route_context = prepare_route_context(role_name);

            if (ev.type == NetCfgChangeType::ROUTE_ADDED)
            {
                AWS::Route::replace_create_route(*route_context,
                                                 route_table_id,
                                                 cidr,
                                                 AWS::Route::RouteTargetType::INSTANCE_ID,
                                                 route_context->instance_id(),
                                                 ipv6);

                vpcRoutes.emplace(cidr, ipv6);

                log_sender.LogInfo("Added route " + cidr);
            }
            else
            {
                AWS::Route::delete_route(*route_context,
                                         route_table_id,
                                         cidr,
                                         ipv6);

                auto it = vpcRoutes.find(VpcRoute(cidr, ipv6));
                if (it != vpcRoutes.end())
                {
                    vpcRoutes.erase(it);
                }

                log_sender.LogInfo("Removed route " + cidr);
            }
        }
        catch (const std::exception& ex)
        {
            log_sender.LogError("Error updating VPC routing: " + std::string(ex.what()));
        }
    }


    GVariant * callback_get_property(GDBusConnection *conn,
                                     const std::string sender,
                                     const std::string obj_path,
                                     const std::string intf_name,
                                     const std::string property_name,
                                     GError **error) override
    {
        GVariant *ret = nullptr;

        ret = nullptr;
        g_set_error (error,
                     G_IO_ERROR,
                     G_IO_ERROR_FAILED,
                     "Unknown property");
        return ret;
    }


    GVariantBuilder * callback_set_property(GDBusConnection *conn,
                                            const std::string sender,
                                            const std::string obj_path,
                                            const std::string intf_name,
                                            const std::string property_name,
                                            GVariant *value,
                                            GError **error) override
    {
        THROW_DBUSEXCEPTION("AWSDBus", "set property not implemented");
    }


    void callback_method_call(GDBusConnection *conn,
                              const std::string sender,
                              const std::string obj_path,
                              const std::string intf_name,
                              const std::string method_name,
                              GVariant *params,
                              GDBusMethodInvocation *invoc) override
    {
    }


private:
    NetCfgProxy::Manager netcfgmgr;
    std::string role_name;
    std::string network_interface_id;
    std::string route_table_id;
    typedef std::pair<std::string, bool> VpcRoute;
    std::set<VpcRoute> vpcRoutes;
    LogSender log_sender;

    std::string read_role_name(const std::string& config_file)
    {
        try
        {
            auto json_content = json::read_fast(config_file);
            return json::get_string(json_content, "role");
        }
        catch (const std::exception & ex)
        {
            log_sender.LogError("Error reading role name: " + std::string(ex.what()));
            return "";
        }
    }

    std::unique_ptr<AWS::Route::Context> prepare_route_context(const std::string& role_name)
    {
        RandomAPI::Ptr rng(new SSLLib::RandomAPI(false));
        AWS::PCQuery::Info ii;

        WS::ClientSet::run_synchronous([&](WS::ClientSet::Ptr cs) {
            AWS::PCQuery::Ptr awspc(new AWS::PCQuery(cs, role_name));
            awspc->start([&](AWS::PCQuery::Info info) {
                ii = std::move(info);
            });
        }, nullptr, rng.get());

        return std::unique_ptr<AWS::Route::Context>(new AWS::Route::Context(ii, ii.creds, rng, nullptr, 0));
    }
};


class AWSDBus : public DBus
{
public:
    explicit AWSDBus(GDBusConnection* conn, const std::string& config_file,
                     LogWriter* logwr, unsigned int log_level, bool sig_brdc)
        : DBus(conn,
               OpenVPN3DBus_interf_aws,
               OpenVPN3DBus_path_aws,
               OpenVPN3DBus_interf_aws),
           config_file{config_file},
           logwr{logwr},
           log_level{log_level},
           signal_broadcast{sig_brdc}
    {
    }

    /**
     *  This callback is called when the service was successfully registered
     *  on the D-Bus.
     */
    void callback_bus_acquired() override
    {
        auto root_path = GetRootPath();
        if (root_path.empty())
            THROW_DBUSEXCEPTION("AWSDBus",
                                "callback_bus_acquired() - "
                                "empty root_path for AWSObject, consider upgrading glib");
        aws.reset(new AWSObject(GetConnection(), root_path, config_file,
                                logwr, log_level, signal_broadcast));
        aws->RegisterObject(GetConnection());
    }


    /**
     *  This is called each time the well-known bus name is successfully
     *  acquired on the D-Bus.
     *
     *  This is not used, as the preparations already happens in
     *  callback_bus_acquired()
     *
     * @param conn     Connection where this event happened
     * @param busname  A string of the acquired bus name
     */
    void callback_name_acquired(GDBusConnection *conn, std::string busname) override
    {
    }

    /**
     *  This is called each time the well-known bus name is removed from the
     *  D-Bus.  In our case, we just throw an exception and starts shutting
     *  down.
     *
     * @param conn     Connection where this event happened
     * @param busname  A string of the lost bus name
     */
    void callback_name_lost(GDBusConnection *conn, std::string busname) override
    {
        THROW_DBUSEXCEPTION("AWSDBus",
                            "openvpn3-service-aws could not register '"
                            + busname + "' on the D-Bus");
    }

private:
    std::string config_file;
    LogWriter* logwr;
    unsigned int log_level;
    bool signal_broadcast;
    AWSObject::Ptr aws;
};


int aws_main(ParsedArgs::Ptr args)
{
    // This program does not require root privileges,
    // so if used - drop those privileges
    drop_root();

    //
    // Open a log destination, if requested
    //
    // This is opened before dropping privileges, to more easily tackle
    // scenarios where logging goes to a file in /var/log or other
    // directories where only root has access
    //
    std::ofstream logfs;
    std::ostream  *logfile = nullptr;
    LogWriter::Ptr logwr = nullptr;
    ColourEngine::Ptr colourengine = nullptr;

    if (args->Present("log-file"))
    {
        std::string fname = args->GetValue("log-file", 0);

        if ("stdout:" != fname)
        {
            logfs.open(fname.c_str(), std::ios_base::app);
            logfile = &logfs;
        }
        else
        {
            logfile = &std::cout;
        }

        if (args->Present("colour"))
        {
            colourengine.reset(new ANSIColours());
            logwr.reset(new ColourStreamWriter(*logfile,
                                               colourengine.get()));
        }
        else
        {
            logwr.reset(new StreamLogWriter(*logfile));
        }
    }

    DBus dbus(G_BUS_TYPE_SYSTEM);
    dbus.Connect();

    unsigned int log_level = 3;
    if (args->Present("log-level"))
    {
        log_level = std::atoi(args->GetValue("log-level", 0).c_str());
    }

    // Initialize logging in the OpenVPN 3 Core library
    openvpn::CoreDBusLogBase corelog(dbus.GetConnection(),
                                     OpenVPN3DBus_interf_aws + ".core",
                                     LogGroup::EXTSERVICE,
                                     logwr.get());
    corelog.SetLogLevel(log_level);


    bool signal_broadcast = args->Present("signal-broadcast");
    LogServiceProxy::Ptr logsrvprx = nullptr;
    if (!signal_broadcast)
    {
        try
        {
            LogServiceProxy::Ptr log_service(new LogServiceProxy(dbus.GetConnection()));
            log_service->Attach(OpenVPN3DBus_interf_aws);
            log_service->Attach(OpenVPN3DBus_interf_aws + ".core");
        }
        catch (const DBusProxyAccessDeniedException& exc)
        {
            std::stringstream err;
            err << "** ERROR **  Could not attach to OpenVPN 3 log service"
                << " (" << std::string(exc.what()) << ")";
            throw CommandException("AWS", err.str());
        }
    }

    std::string config_file{OPENVPN3_AWS_CONFIG};
    if (args->Present("config"))
    {
        config_file = args->GetValue("config", 0);
    }

    // Initialize Core library
    InitProcess::init();

    // Prepare the GLib GMainLoop
    GMainLoop *main_loop = g_main_loop_new(nullptr, FALSE);
    g_unix_signal_add(SIGINT, stop_handler, main_loop);
    g_unix_signal_add(SIGTERM, stop_handler, main_loop);

    try
    {
        // Prepare the net.openvpn.v3.aws D-Bus service
        AWSDBus aws(dbus.GetConnection(), config_file,
                    logwr.get(), log_level, signal_broadcast);
        aws.Setup();

        // Start the main program loop
        g_main_loop_run(main_loop);

        // Clean up before exit
        g_main_loop_unref(main_loop);
        if (logsrvprx)
        {
            logsrvprx->Detach(OpenVPN3DBus_interf_backends);
        }
        InitProcess::uninit();
    }
    catch (const DBusException& exc)
    {
        throw CommandException("AWS", exc.what());
    }

    return 0;
}


int main(int argc, char **argv)
{
    SingleCommand cmd(argv[0], "OpenVPN 3 AWS VPC integration service",
                             aws_main);
    cmd.AddVersionOption();
    cmd.AddOption("config", 'c', "FILE", true,
                  "AWS VPC configuration file (default: " OPENVPN3_AWS_CONFIG ")");
    cmd.AddOption("log-file", "FILE" , true,
                  "Write log data to FILE.  Use 'stdout:' for console logging.");
    cmd.AddOption("log-level", "LOG-LEVEL", true,
                  "Log verbosity level (valid values 0-6, default 3)");
    cmd.AddOption("colour", 0,
                        "Make the log lines colourful");
    cmd.AddOption("signal-broadcast", 0,
                  "Broadcast all D-Bus signals from openvpn3-service-aws");
    try
    {
        return cmd.RunCommand(simple_basename(argv[0]), argc, argv);
    }
    catch (CommandException& excp)
    {
        std::cout << cmd.GetCommand() << ": " << excp.what() << std::endl;
        return 2;
    }

}
